### DISCUSSION TITLE
Why gRPC used in basket api?

### DISCUSSION THREAD

**Discussion by ks1990cn** [dotnet/eShop#167] [Q&A]
Hi, 
Can anyone explain why basket api using GRPC, any specific reason?

**Comment by halilkocaoz** [comment]
Basket API takes a lot of requests than other services because of user interactions like adding new items, updating, checking prices and GRPC uses Protobuf which is more efficient and faster than JSON or XML used in other APIs, with GRPC you can handle requests more faster than REST.

Basket API needs real time handling for inventory management, price updates, concurrent handling to prevent issues like double booking or selling more items than are in stock with GRPC bi-directional streaming useful for real time functionality. The Basket API also requires real-time handling so that changes made on one device are immediately reflected on all other devices.

**Reply by ks1990cn** [reply]
Are you assuming more traffic on basket api than catalog api? Interesting, why?

**Reply by ks1990cn** [reply]
Catalog service may have more traffic I guess cause users will be exploring, basket service will make updates on catalog after purchase is done.
Business owners will be frequently working and updating catalog service.

Basket api looks like just used mostly by person who wants to buy anything.

**Reply by halilkocaoz** [reply]
You are right Catalog service may have more traffic but it doesn't need to be real time response.

**Reply by ks1990cn** [reply]
You are adding just to redis, how it is real time? You are not making chat api there!!!
What if your caching server goes slow? Will you still have realtime?😑

If you setup your payment APIs, all api you will select gRPC?😑

Basketapi will have to maintain consistency with price by event. If catalog service have price changed basket should have changed.

**Reply by ks1990cn** [reply]
@davidfowl @captainsafia we need expert answer, please help us!!
🆘 Discuss mode😛

**Reply by ks1990cn** [reply]
Earlier eshoponcontainer PDF book discuss about basket as an API, but now we can see this as gRPC. Hope any book on this going ahead will be updated with same reason..

On PDF book, there is no mention of grpc in it.

https://learn.microsoft.com/en-us/dotnet/architecture/microservices/

**Reply by ks1990cn** [reply]
https://aws.amazon.com/blogs/mobile/backends-for-frontends-pattern/

Is there to create realtime update to client using websockets @halilkocaoz , not basket api directly.
